File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/sgcs/SGCManager.kt
package com.mentra.bluetoothsdk.sgcs

import com.mentra.bluetoothsdk.Bridge
import com.mentra.bluetoothsdk.DeviceStore
import com.mentra.bluetoothsdk.utils.ConnTypes

abstract class SGCManager {
    // Hard coded device properties:
    @JvmField var type: String = ""
    @JvmField var hasMic: Boolean = false

    // Audio Control
    abstract fun setMicEnabled(enabled: Boolean)
    abstract fun sortMicRanking(list: MutableList<String>): MutableList<String>

    // Camera & Media
    abstract fun requestPhoto(
            requestId: String,
            appId: String,
            size: String,
            webhookUrl: String?,
            authToken: String?,
            compress: String?,
            flash: Boolean,
            sound: Boolean,
            exposureTimeNs: Long?,
    )
    abstract fun startStream(message: MutableMap<String, Any>)
    abstract fun stopStream()
    abstract fun sendStreamKeepAlive(message: MutableMap<String, Any>)
    abstract fun startVideoRecording(requestId: String, save: Boolean, flash: Boolean, sound: Boolean)
    abstract fun stopVideoRecording(requestId: String)

    // Button Settings
    abstract fun sendButtonPhotoSettings()
    abstract fun sendButtonVideoRecordingSettings()
    abstract fun sendButtonMaxRecordingTime()
    abstract fun sendButtonCameraLedSetting()
    abstract fun sendCameraFovSetting()

    // Display Control
    abstract fun setBrightness(level: Int, autoMode: Boolean)
    abstract fun clearDisplay()
    abstract fun sendTextWall(text: String)
    abstract fun sendDoubleTextWall(top: String, bottom: String)
    abstract fun displayBitmap(base64ImageData: String): Boolean
    abstract fun showDashboard()
    abstract fun setDashboardPosition(height: Int, depth: Int)

    /** Default: full [setDashboardPosition] (e.g. G1 single command). Nex overrides to height protobuf only. */
    open fun setDashboardHeightOnly(height: Int) {
        val depth = (DeviceStore.store.get("bluetooth", "dashboard_depth") as? Number)?.toInt() ?: 2
        setDashboardPosition(height, depth)
    }

    /** Default: full [setDashboardPosition]. Nex overrides to display_distance only. */
    open fun setDashboardDepthOnly(depth: Int) {
        val height = (DeviceStore.store.get("bluetooth", "dashboard_height") as? Number)?.toInt() ?: 4
        setDashboardPosition(height, depth)
    }

    // Dashboard Menu (default no-op — only G2 supports this)
    open fun setDashboardMenu(items: List<Map<String, Any>>) {}

    // Controller bridging (default no-op — only G2 supports pairing with a ring controller)
    open fun connectController() {}
    open fun disconnectController() {}

    // Device Control
    abstract fun setHeadUpAngle(angle: Int)
    abstract fun getBatteryStatus()
    abstract fun setSilentMode(enabled: Boolean)
    abstract fun exit()
    abstract fun sendShutdown()
    abstract fun sendReboot()
    abstract fun sendRgbLedControl(
            requestId: String,
            packageName: String?,
            action: String,
            color: String?,
            onDurationMs: Int,
            offDurationMs: Int,
            count: Int
    )

    // Connection Management
    abstract fun disconnect()
    abstract fun forget()
    abstract fun findCompatibleDevices()
    abstract fun stopScan()
    abstract fun connectById(id: String)
    abstract fun getConnectedBluetoothName(): String
    abstract fun cleanup()
    abstract fun ping()
    abstract fun dbg1()
    abstract fun dbg2()

    // Network Management
    abstract fun requestWifiScan()
    abstract fun sendWifiCredentials(ssid: String, password: String)
    abstract fun forgetWifiNetwork(ssid: String)
    abstract fun sendHotspotState(enabled: Boolean)

    /** Set glasses system clock (Mentra Live only; no-op on other devices). */
    open fun sendSetSystemTime(timestampMs: Long) {
        Bridge.log("SGC: sendSetSystemTime not supported on $type")
    }

    // User Context (for crash reporting)
    abstract fun sendUserEmailToGlasses(email: String)

    // Incident Reporting
    abstract fun sendIncidentId(incidentId: String, apiBaseUrl: String? = null)

    // Gallery
    abstract fun queryGalleryStatus()
    abstract fun sendGalleryMode()

    // Voice Activity Detection
    open fun sendVoiceActivityDetectionSetting() {}

    // Version info
    abstract fun requestVersionInfo()

    // DeviceStore-backed read-only getters for convenience
    val fullyBooted: Boolean
        get() = DeviceStore.get("glasses", "fullyBooted") as? Boolean ?: false

    val connected: Boolean
        get() = DeviceStore.get("glasses", "connected") as? Boolean ?: false

    val connectionState: String
        get() = DeviceStore.get("glasses", "connectionState") as? String ?: ConnTypes.DISCONNECTED

    val appVersion: String
        get() = DeviceStore.get("glasses", "appVersion") as? String ?: ""

    val buildNumber: String
        get() = DeviceStore.get("glasses", "buildNumber") as? String ?: ""

    val deviceModel: String
        get() = DeviceStore.get("glasses", "deviceModel") as? String ?: ""

    val androidVersion: String
        get() = DeviceStore.get("glasses", "androidVersion") as? String ?: ""

    val otaVersionUrl: String
        get() = DeviceStore.get("glasses", "otaVersionUrl") as? String ?: ""

    val firmwareVersion: String
        get() = DeviceStore.get("glasses", "firmwareVersion") as? String ?: ""

    val bluetoothMacAddress: String
        get() = DeviceStore.get("glasses", "bluetoothMacAddress") as? String ?: ""

    val serialNumber: String
        get() = DeviceStore.get("glasses", "serialNumber") as? String ?: ""

    val style: String
        get() = DeviceStore.get("glasses", "style") as? String ?: ""

    val color: String
        get() = DeviceStore.get("glasses", "color") as? String ?: ""

    val micEnabled: Boolean
        get() = DeviceStore.get("glasses", "micEnabled") as? Boolean ?: false

    val voiceActivityDetectionEnabled: Boolean
        get() = DeviceStore.get("glasses", "voiceActivityDetectionEnabled") as? Boolean ?: true

    val batteryLevel: Int
        get() = DeviceStore.get("glasses", "batteryLevel") as? Int ?: -1

    val headUp: Boolean
        get() = DeviceStore.get("glasses", "headUp") as? Boolean ?: false

    val charging: Boolean
        get() = DeviceStore.get("glasses", "charging") as? Boolean ?: false

    val caseOpen: Boolean
        get() = DeviceStore.get("glasses", "caseOpen") as? Boolean ?: true

    val caseRemoved: Boolean
        get() = DeviceStore.get("glasses", "caseRemoved") as? Boolean ?: true

    val caseCharging: Boolean
        get() = DeviceStore.get("glasses", "caseCharging") as? Boolean ?: false

    val caseBatteryLevel: Int
        get() = DeviceStore.get("glasses", "caseBatteryLevel") as? Int ?: -1

    val wifiSsid: String
        get() = DeviceStore.get("glasses", "wifiSsid") as? String ?: ""

    val wifiConnected: Boolean
        get() = DeviceStore.get("glasses", "wifiConnected") as? Boolean ?: false

    val wifiLocalIp: String
        get() = DeviceStore.get("glasses", "wifiLocalIp") as? String ?: ""

    val hotspotEnabled: Boolean
        get() = DeviceStore.get("glasses", "hotspotEnabled") as? Boolean ?: false

    val hotspotSsid: String
        get() = DeviceStore.get("glasses", "hotspotSsid") as? String ?: ""

    val hotspotPassword: String
        get() = DeviceStore.get("glasses", "hotspotPassword") as? String ?: ""

    val hotspotGatewayIp: String
        get() = DeviceStore.get("glasses", "hotspotGatewayIp") as? String ?: ""
}

File: mobile/modules/bluetooth-sdk/ios/Source/sgcs/SGCManager.swift
import Foundation

@MainActor
protocol SGCManager {
    // MARK: - hard coded device properties:

    var type: String { get set }
    var hasMic: Bool { get }

    // MARK: - Audio Control

    func setMicEnabled(_ enabled: Bool)
    func sortMicRanking(list: [String]) -> [String]

    // MARK: - Messaging

    func sendJson(_ jsonOriginal: [String: Any], wakeUp: Bool, requireAck: Bool)

    // MARK: - Camera & Media

    func requestPhoto(
        _ requestId: String, appId: String, size: String?, webhookUrl: String?, authToken: String?,
        compress: String?, flash: Bool, sound: Bool, exposureTimeNs: Double?
    )
    func startStream(_ message: [String: Any])
    func stopStream()
    func sendStreamKeepAlive(_ message: [String: Any])
    func startVideoRecording(requestId: String, save: Bool, flash: Bool, sound: Bool)
    func stopVideoRecording(requestId: String)

    // MARK: - Button Settings

    func sendButtonPhotoSettings()
    func sendButtonVideoRecordingSettings()
    func sendButtonMaxRecordingTime()
    func sendButtonCameraLedSetting()
    func sendCameraFovSetting()

    // MARK: - Display Control

    func setBrightness(_ level: Int, autoMode: Bool)
    func clearDisplay()
    func sendTextWall(_ text: String)
    func sendDoubleTextWall(_ top: String, _ bottom: String)
    func displayBitmap(base64ImageData: String) async -> Bool
    func showDashboard()
    func setDashboardPosition(_ height: Int, _ depth: Int)
    /// Default implementation sends both via [setDashboardPosition]; Nex overrides to one protobuf.
    func setDashboardHeightOnly(_ height: Int)
    func setDashboardDepthOnly(_ depth: Int)

    // MARK: - Dashboard Menu

    func setDashboardMenu(_ items: [[String: Any]])

    // MARK: - Device Control

    func setHeadUpAngle(_ angle: Int)
    func getBatteryStatus()
    func setSilentMode(_ enabled: Bool)
    func exit()
    func sendShutdown()
    func sendReboot()
    func sendRgbLedControl(
        requestId: String, packageName: String?, action: String, color: String?, onDurationMs: Int,
        offDurationMs: Int, count: Int
    )

    // MARK: - Connection Management

    func disconnect()
    func forget()
    func findCompatibleDevices()
    func stopScan()
    func connectById(_ id: String)
    func getConnectedBluetoothName() -> String?
    func connectController()
    func disconnectController()
    func cleanup()
    func ping()
    func dbg1()
    func dbg2()

    // MARK: - Network Management

    func requestWifiScan()
    func sendWifiCredentials(_ ssid: String, _ password: String)
    func forgetWifiNetwork(_ ssid: String)
    func sendHotspotState(_ enabled: Bool)
    func sendOtaStart()
    func sendOtaQueryStatus()
    func sendSetSystemTime(_ timestampMs: Int64)
    func sendOtaRetryVersionCheck()

    // MARK: - User Context (for crash reporting)

    func sendUserEmailToGlasses(_ email: String)

    // MARK: - Incident Reporting

    func sendIncidentId(_ incidentId: String, apiBaseUrl: String?)

    // MARK: - Gallery

    func queryGalleryStatus()
    func sendGalleryMode()

    // MARK: - Voice Activity Detection

    func sendVoiceActivityDetectionSetting()

    // MARK: - Version Info

    func requestVersionInfo()
}

/// doesn't seem to work for concurrency reasons :(
/// we can make read-only getters for convienence though:
extension SGCManager {
    // MARK: - Dashboard (default: combined wire format; Nex implements single-field)

    func setDashboardHeightOnly(_ height: Int) {
        let d = DeviceStore.shared.get("bluetooth", "dashboard_depth") as? Int ?? 2
        setDashboardPosition(height, d)
    }

    func setDashboardDepthOnly(_ depth: Int) {
        let h = DeviceStore.shared.get("bluetooth", "dashboard_height") as? Int ?? 4
        setDashboardPosition(h, depth)
    }

    // MARK: - Dashboard Menu (default no-op — only G2 supports this)

    func setDashboardMenu(_: [[String: Any]]) {}

    // MARK: - Voice Activity Detection (default no-op — Mentra Live supports this)

    func sendVoiceActivityDetectionSetting() {}

    /// Default no-op; Mentra Live overrides when phone detects clock skew during gallery sync.
    func sendSetSystemTime(_: Int64) {
        Bridge.log("SGC: sendSetSystemTime not supported")
    }

    func sendOtaRetryVersionCheck() {
        Bridge.log("SGC: sendOtaRetryVersionCheck not supported")
    }

    // MARK: - Default DeviceStore-backed property implementations

    var fullyBooted: Bool {
        DeviceStore.shared.get("glasses", "fullyBooted") as? Bool ?? false
    }

    var connected: Bool {
        DeviceStore.shared.get("glasses", "connected") as? Bool ?? false
    }

    var appVersion: String {
        DeviceStore.shared.get("glasses", "appVersion") as? String ?? ""
    }

    var buildNumber: String {
        DeviceStore.shared.get("glasses", "buildNumber") as? String ?? ""
    }

    var deviceModel: String {
        DeviceStore.shared.get("glasses", "deviceModel") as? String ?? ""
    }

    var androidVersion: String {
        DeviceStore.shared.get("glasses", "androidVersion") as? String ?? ""
    }

    var otaVersionUrl: String {
        DeviceStore.shared.get("glasses", "otaVersionUrl") as? String ?? ""
    }

    var firmwareVersion: String {
        DeviceStore.shared.get("glasses", "firmwareVersion") as? String ?? ""
    }

    var bluetoothMacAddress: String {
        DeviceStore.shared.get("glasses", "bluetoothMacAddress") as? String ?? ""
    }

    var serialNumber: String {
        DeviceStore.shared.get("glasses", "serialNumber") as? String ?? ""
    }

    var style: String {
        DeviceStore.shared.get("glasses", "style") as? String ?? ""
    }

    var color: String {
        DeviceStore.shared.get("glasses", "color") as? String ?? ""
    }

    var micEnabled: Bool {
        DeviceStore.shared.get("glasses", "micEnabled") as? Bool ?? false
    }

    var voiceActivityDetectionEnabled: Bool {
        DeviceStore.shared.get("glasses", "voiceActivityDetectionEnabled") as? Bool ?? true
    }

    var batteryLevel: Int {
        DeviceStore.shared.get("glasses", "batteryLevel") as? Int ?? -1
    }

    var headUp: Bool {
        DeviceStore.shared.get("glasses", "headUp") as? Bool ?? false
    }

    var charging: Bool {
        DeviceStore.shared.get("glasses", "charging") as? Bool ?? false
    }

    var caseOpen: Bool {
        DeviceStore.shared.get("glasses", "caseOpen") as? Bool ?? true
    }

    var caseRemoved: Bool {
        DeviceStore.shared.get("glasses", "caseRemoved") as? Bool ?? true
    }

    var caseCharging: Bool {
        DeviceStore.shared.get("glasses", "caseCharging") as? Bool ?? false
    }

    var caseBatteryLevel: Int {
        DeviceStore.shared.get("glasses", "caseBatteryLevel") as? Int ?? -1
    }

    var wifiSsid: String {
        DeviceStore.shared.get("glasses", "wifiSsid") as? String ?? ""
    }

    var wifiConnected: Bool {
        DeviceStore.shared.get("glasses", "wifiConnected") as? Bool ?? false
    }

    var wifiLocalIp: String {
        DeviceStore.shared.get("glasses", "wifiLocalIp") as? String ?? ""
    }

    var hotspotEnabled: Bool {
        DeviceStore.shared.get("glasses", "hotspotEnabled") as? Bool ?? false
    }

    var hotspotSsid: String {
        DeviceStore.shared.get("glasses", "hotspotSsid") as? String ?? ""
    }

    var hotspotPassword: String {
        DeviceStore.shared.get("glasses", "hotspotPassword") as? String ?? ""
    }

    var hotspotGatewayIp: String {
        DeviceStore.shared.get("glasses", "hotspotGatewayIp") as? String ?? ""
    }
}
